apps/prairielearn/src/middlewares/clientFingerprint.ts

```diff
@@ -0,0 +1,42 @@
+import asyncHandler from 'express-async-handler';
+import * as sqldb from '@prairielearn/postgres';
+
+const sql = sqldb.loadSqlEquiv(__filename);
+
+export default asyncHandler(async (req, res, next) => {
+  const params = {
+    ip_address: req.ip,
+    user_id: res.locals.user.user_id,
+    session_id: req.session.id,
+    user_agent: req.headers['user-agent'],
+    accept_language: req.headers['accept-language'],
+    accept: req.headers['accept'],
+  };
+  // Need a better way to find or insert a fingerprint. Currently doing this through two calls but would probably be better to do it in one.
+  // First call, check if fingerprint exists in the database
+  let client_fingerprint_id: number | null =
+    parseInt(
+      (await sqldb.queryZeroOrOneRowAsync(sql.select_client_fingerprint, params)).rows[0]?.id,
+    ) ?? null;
+  //if the fingerprint doesn't exist, create a new fingerprint
+  if (!client_fingerprint_id) {
+    client_fingerprint_id = (await sqldb.queryOneRowAsync(sql.insert_client_fingerprint, params))
+      .rows[0].id;
+  }
```

Comment:
Will delete the commented section after resolving. Is there a cleaner way of finding a fingerprint if it exists and if not, inserting it? I believe in our originally meeting we discussed something similar to an upsert here but I am not sure the best way to update the SQL query for that.

Reply:
I'll defer to @mwest1066 here, who is the resident expert at being clever with Postgres!

Reply:
I believe you will need an index for this search. If you make this index unique (which I think you should), you could use an `INSERT ... ON CONFLICT`, but this may be an issue:
* `ON CONFLICT DO NOTHING` is not an option since it would not return the ID in case of conflict.
* `ON CONFLICT DO UPDATE` could work, if there is something you want to update (date of last use, or something similar?), but would not be an option if there's nothing to update.

Reply:
After discussion with @nwalters512 and @mwest1066, we have decided it is best stick with two queries in this scenario.

Reply:
We explicitly discussed this and decided that the risk of a race condition is tolerable given how unlikely it is.